### PR TITLE
Bug 1564811 - Badge actions can define a timeout delay after which the action is executed

### PR DESCRIPTION
**Pull request by piatra**

### CODE REVIEW

File: lib/OnboardingMessageProvider.jsm [hunk 1]
```diff
@@ -480,6 +480,8 @@ const ONBOARDING_MESSAGES = async () => [
     id: `WHATS_NEW_BADGE_${FIREFOX_VERSION}`,
     template: "toolbar_badge",
     content: {
+      // delay: 5 * 3600 * 1000,
+      delay: 5000,
```
Comment (piatra) [inline lib/OnboardingMessageProvider.jsm]: Just for testing, will revert after review.

File: lib/ASRouter.jsm [hunk 2]
```diff
@@ -732,11 +732,6 @@ class _ASRouter {
     ToolbarPanelHub.init({
       getMessages: this.handleMessageRequest,
     });
-    ToolbarBadgeHub.init(this.waitForInitialized, {
-      handleMessageRequest: this.handleMessageRequest,
-      addImpression: this.addImpression,
-      blockMessageById: this.blockMessageById,
-    });
```
Comment (rlr) [inline lib/ASRouter.jsm]: Oh this was in there twice? oops, probably a bad merge conflict.

File: lib/ToolbarBadgeHub.jsm [hunk 3]
```diff
@@ -76,7 +95,9 @@ class _ToolbarBadgeHub {
         .querySelector(".toolbarbutton-badge")
         .setAttribute("value", "x");
 
-      toolbarbutton.addEventListener("click", this.removeAllNotifications, {
+      // `mousedown` event required because of the `onmousedown` defined on
```
[the remark this answers is not in the document]
Comment (rlr) [reply lib/ToolbarBadgeHub.jsm]: oh actually. I think listening to `click` will handle all the keyboard access cases. So just listen for `click` and `mousedown` maybe?

[the remark this answers is not in the document]
Comment (rlr) [reply lib/ToolbarBadgeHub.jsm]: Gah. Another edge case: if you right click with mouse, the notification goes away but the panel doesn't open.

File: lib/OnboardingMessageProvider.jsm (hunk 1 above)
Comment (rlr) [reply lib/OnboardingMessageProvider.jsm]: Hmm, when you put the old delay back in, it's going to be hard to test from devtools. We'll need to find a workaround probably.

Comment (piatra) [reply lib/OnboardingMessageProvider.jsm]: I was thinking we move all our test messages to staging, it's easy to switch using the RS-devtools and we can set values that make it easier to test.

**Comment by rlr** [comment]
I cant tell what Taskcluster is unhappy about... I see this `[36m 1:11.40(B[m [34mINFO(B[m Console message: [JavaScript Error: "getScreenshot(https://bugzilla.mozilla.org) failed: TypeError: NetworkError when attempting to fetch resource." {file: "resource://activity-stream/lib/Screenshots.jsm" line: 59}]` which can't be this patch.

**Comment by Mardak** [comment]
Looks like it timed out on matchPattern:
```
INFO Leaving test bound test_onLocationChange_cb
INFO Entering test bound test_matchPattern
```

Although running locally with `./mach test asrouter_cfr --verify --headless` seems to be fine for me

**Comment by Mardak** [comment]
Ah looks intermittent. The retrigger is already past asrouter_cfr

**Comment by piatra** [comment]
Same, I wasn't able to reproduce locally.

**Comment by Mardak** [comment]
There was a good run. good to merge! 😉

**Pull request merged**